docker/daemon/daemon_dest.go

```diff
@@ -157,15 +157,26 @@ func (d *daemonImageDestination) PutBlob(stream io.Reader, inputInfo types.BlobI
 		defer os.Remove(streamCopy.Name())
 		defer streamCopy.Close()
 
-		size, err := io.Copy(streamCopy, stream)
+		// If we need to compute a digest.
+		digester := digest.Canonical.New()
+		tee := io.TeeReader(stream, digester.Hash())
+
+		size, err := io.Copy(streamCopy, tee)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
 		_, err = streamCopy.Seek(0, os.SEEK_SET)
 		if err != nil {
 			return types.BlobInfo{}, err
 		}
+
+		if inputInfo.Digest == "" {
+			// In some (hacky) cases a caller might not know the size, but will
+			// know the digest (where "know" == "have a fake digest").
+			inputInfo.Digest = digester.Digest()
```

Comment:
Even if it is broken, can we **please** fix it separately. I think the reason this is done this way is because of the whole shitty DiffID code -- which I don't think we can completely fix anyway. I'm just trying to make this code work.